---
title: "stemi_link_peakckmb_to_scrnaseq"
author: "Roy Oelen"
date: "2023-01-02"
output: html_document
---

```{r header, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
#!/usr/bin/env Rscript
############################################################################################################################
# Authors: Roy Oelen
# Name: stemi_link_peakckmb_to_scrnaseq.Rmd
# Function:
############################################################################################################################
```

```{r libraries, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
####################
# libraries        #
####################

# for loading scRNAseq data
library(Seurat)
# for differential expression
library(MAST)
# to perform parallel operations
library(foreach)
library(doParallel)

```

```{r functions, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
####################
# Functions        #
####################

# calculating peak CK-MB values of every participant
add_peak_ck_mb <- function(clinvar_object){
  # subset to the ckmb variables
  ckmb <- clinvar[, grep('ck_mb', colnames(clinvar_object))]
  # change the NA variables to -1
  ckmb[is.na(ckmb)] <- -1
  # get the maximum
  max_ckmb <- apply(ckmb, 1, max)
  # replace a -1 with NA, this means there was no real variable
  max_ckmb[max_ckmb == -1] <- NA
  # paste onto the existing object
  clinvar_object$peak_ck_mb <- max_ckmb
  return(clinvar_object)
}

calc_age_in_days <- function(date_of_birth_string, split_character='/', day_of_month_col=1, month_of_year_col=2, year_col=3){
  # split the string
  split_string <- strsplit(date_of_birth_string, split = split_character)[[1]]
  day = as.numeric(split_string[day_of_month_col])
  month = as.numeric(split_string[month_of_year_col])
  year = as.numeric(split_string[year_col])
  # create new string as input for the Date library
  standard_data_string <- paste(day, month, year, sep = '/')
  # turn into date object
  birth <- as.Date(standard_data_string, format='%d/%m/%Y')
  # get now
  now_string <- Sys.Date()
  # convert to Date as well
  now <- as.Date(now_string, format='%Y-%m-%d')
  # get the time between now and the birth
  age_days <- difftime(now, birth, units = 'days')
  return(age_days)
}

add_clinical_variables <- function(seurat_object, clinvar_object, clinical_vars_to_add, assignment_column_seurat='assignment.final', assignment_column_clinical='record_id', assignment_column_clinical_prepend=''){
  # subset the seurat object to just the metadata assignments
  meta.data <- seurat_object@meta.data[, c(assignment_column_seurat), drop = F]
  # make that a character type in case it was a factor
  meta.data[[assignment_column_seurat]] <- as.character(meta.data[[assignment_column_seurat]])
  # construct the assignment in the clinical variable object
  clinvar_object$assignment.seurat <- paste(assignment_column_clinical_prepend, as.character(clinvar_object[[assignment_column_clinical]]), sep = '')
  # subset the metadata to what we could possibly match, to save time
  meta.data <- meta.data[meta.data[[assignment_column_seurat]] %in% unique(clinvar_object$assignment.seurat), , drop = F]
  # grab the clinical variables in the order of the metadata by matching the ID column
  clinvar_matched <- clinvar_object[match(meta.data[[assignment_column_seurat]], clinvar_object$assignment.seurat), clinical_vars_to_add]
  # the order is now the same, so we can cbind
  meta.data <- cbind(meta.data, clinvar_matched)
  # now we can add it to the object, by subsetting the metadata by columns, it'll match by the rownames(cell barcodes)
  seurat_object <-AddMetaData(seurat_object, meta.data[, clinical_vars_to_add])
  return(seurat_object)
}

# do the linear regression
de_glmer.func <- function(sca_object, contrast, fixed_effects, random_effects, nagq0=F, alt_order=F){
  # extract the contrast levels
  contrast_levels <- unique(colData(sca_object)[[contrast]])
  # sort alphabetically
  contrast_levels <- contrast_levels[order(contrast_levels)]
  print(paste0('Testing: ', contrast))
  # paste the random effects
  random_effects.fmla <- paste(paste0('(1|',random_effects,')'),collapse='+') #try other nomenclature (=option 1, default)
  # paste the fixed effects
  contrast_fixed.fmla <- paste(c(contrast,fixed_effects),collapse='+')
  # if the fixed effects come before the contrasting effect
  if(alt_order){
    contrast_fixed.fmla <- paste(c(fixed_effects,contrast),collapse='+')
  }
  # paste fixend and random together
  zlm_vars <- paste0('~',paste(c(contrast_fixed.fmla,random_effects.fmla), collapse='+'))
  # turn into a zero-inflated formula with glmer
  zlm_formula <- as.formula(zlm_vars)
  print(paste0('Fitting glmer: ',zlm_vars))
  # an extra parameter can be supplied to stop the convergement failing
  if(nagq0){
    print(paste0('Trying to mitigate model failing to converge for some genes by passing nAGQ=0 to the fitting function zlm (fitArgsD=list(nAGQ=0))'))
    zlmCond <- zlm(zlm_formula,
                   sca_object,
                   method='glmer', ebayes=FALSE, fitArgsD=list(nAGQ=0))
    summaryCond <- summary(zlmCond, doLRT=paste0(contrast), fitArgsD=list(nAGQ=0))
  }else{
    zlmCond <- zlm(zlm_formula,
                   sca_object,
                   method='glmer', ebayes=FALSE)
    summaryCond <- summary(zlmCond, doLRT=paste0(contrast))
  }
  # grab the summary
  summaryDt <- summaryCond$datatable
  # calculate the LFC
  z1_lfc <- getLogFC(dt)
  # save the results
  z1_list <- list(zlm=zlmCond, dt=summaryDt, lfc=z1_lfc)
  # also try a fixed effects formula
  fixed_effects.fmla <- paste0('~',paste(c(contrast, fixed_effects),collapse='+'))
  fixed_effects.formula <- as.formula(fixed_effects.fmla)
  print(paste0('Fitting bayesglm with discrete residuals: ',fixed_effects.fmla))
  z2 <- NULL
  z2_summaryDt <- NULL
  z2_lfc <- NULL
  # depending on this parameter we can or cannot do bayesianglm
  if(nagq0){
    print('bayesian model is incompatible with nagq0 parameter')
  }
  else{
    z2 <- zlm(fixed_effects.formula,
              sca_object,
              method = 'bayesglm',
              hook=discrete_residuals_hook)
    z2_summaryDt <- summary(z2, doLRT=paste0(contrast))$datatable
    z2_lfc <- getLogFC(z2_summaryDt)
  }
  print('residual collection is incompatible with bayesglm + discrete residuals')
  z2_residuals <- NA
  z2_list <- list(zlm=z2, dt=z2_summaryDt, residuals=z2_residuals, lfc=z2_lfc)
  print(paste0('Fitting bayesglm with continuous residuals: ',fixed_effects.fmla))
  z3 <- NULL
  z3_summaryDt <- NULL
  z3_lfc <- NULL
  print('z3')
  if(nagq0){
    print('bayesian model is incompatible with nagq0 parameter + continuous residuals')
  }
  else{
    z3 <- zlm(fixed_effects.formula, sca_object, hook=continuous_residuals_hook)
    z3_summaryDt <- summary(z3, doLRT=paste0(contrast))$datatable
    z3_lfc <- getLogFC(z3_summaryDt)
  }
  z3_residuals <- NA#z3_residuals <- collectResiduals(z3, sca_object)
  z3_list <- list(zlm=z3, dt=z3_summaryDt, residuals=z3_residuals, lfc=z3_lfc)
  print(paste0('Fitting bayesglm with combined residuals: ',fixed_effects.fmla))
  z4 <- NULL
  z4_summaryDt<- NULL
  z4_lfc <- NULL
  if(nagq0){
    print('bayesian model is incompatible with nagq0 parameter')
  }
  else{
    z4 <- zlm(fixed_effects.formula, sca_object, hook=combined_residuals_hook)
    z4_summaryDt <- summary(z4, doLRT=paste0(contrast))$datatable
    z4_lfc <- getLogFC(z4_summaryDt)
  }
  print('residual collection is incompatible with bayesglm + combined residuals')
  z4_residuals <- NA
  z4_list <- list(zlm=z4, dt=z4_summaryDt, residuals=z4_residuals, lfc=z4_lfc)
  # set the contrast to be a fixed effect and do no random effects
  contrast_fixed.formula <- as.formula(paste0('~',contrast_fixed.fmla))
  print(paste0('Fitting bayesglm with fixed contrast: ',fixed_effects.fmla))
  z5 <- NULL
  z5_summaryDt <- NULL
  z5_lfc <- NULL
  if(nagq0){
    print('bayesian model is incompatible with nagq0 parameter')
  }
  else{
    z5 <- zlm(contrast_fixed.formula, sca_object)
    z5_summaryDt <- summary(z5, doLRT=paste0(contrast))$datatable
    z5_lfc <- getLogFC(z5_summaryDt)
  }
  #partialScore(z5, 'MTL.value')
  z5_list <- list(zlm=z5, dt=z5_summaryDt, lfc=z5_lfc)

  # put it all in the list
  z_list <- list('glmer'=z1_list, 'bayesgml_fixed_discrete'=z2_list, 'bayesgml_fixed_continuous'=z3_list, 'bayesgml_fixed_combined'=z4_list, 'glmer_fixed'=z5_list)
  # and return that as a result
  return(z_list)
}


perform_MAST <- function(sca, contrast_var='inflammation_status', fixed_effects=c('cngeneson','gender','age'), random_effects=c('donor_final','lane'), nagq0=F){
  # apply for each contrast
  de_glmer_list <- de_glmer.func(sca_object = sca,
                          contrast = contrast_var,
                          fixed_effects = fixed_effects,
                          random_effects = random_effects,
                          nagq0 = nagq0)
  return(de_glmer_list)
}

# convert a Seurat object to an SCE object
seurat_object_to_singlecellassay <- function(seurat_object, assay='SCT'){
  # Convert Seurat object to SingleCellExperiment object; and then from SingleCellExperiment object to SingleCellAssay object.
  DefaultAssay(seurat_object) <- assay
  sca <- as(as.SingleCellExperiment(seurat_object), 'SingleCellAssay')
  # Filter out lowly variable genes
  sca <- sca[freq(sca)>0,] #returns the frequency of expression, i.e., the proportion of non-zero values in sc
  colData(sca)$wellKey<-colnames(sca)
  rowData(sca)$primerid <- rownames(rowData(sca))
  colData(sca) <- droplevels(colData(sca))
  cdr2 <- colSums(assay(sca)>0) #assay(sca) is working on assays(sca)$logcounts
  colData(sca)$cngeneson <- scale(cdr2)
  return(sca)
}

do_mast_celltype <- function(seurat_object, cell_type, cell_type_column='cell_type', contrast_var='peak_ck_mb', fixed_effects=c('cngeneson','gender','age', 'timepoint.final'), random_effects=c('assignment.final','lane'), nagq0=F, assay='SCT', freq_expressed=0.1, tsv_loc=NULL, rds_loc=NULL, simplify_result=F, min_lfc=NULL){
  print(paste('analyzing', cell_type))
  # subset to the cell type
  seurat_object_cell_type <- seurat_object[, seurat_object@meta.data[[cell_type_column]] == cell_type]
  # save the genes initially in the object
  cell_type_genes <- rownames(seurat_object_cell_type)
  # subset to an LFC if requested
  if (!is.null(min_lfc)){
    # calculate the average expression
    avg_exps <- data.frame(AverageExpression(seurat_object_cell_type, assays = c(assay), group.by = contrast_var)[[assay]])
    # remove zero expression
    avg_exps <- avg_exps[apply(avg_exps, 1, sum) > 0, ]
    # calculate LFC
    avg_exps[['LFC']] <- log2(avg_exps[[1]] / avg_exps[[2]])
    # and the absolute value
    avg_exps[['abs_LFC']] <- abs(avg_exps[['LFC']])
    # get the genes that fit the LFC
    genes_lfc <- rownames(avg_exps[!is.na(avg_exps[['abs_LFC']]) & is.finite(avg_exps[['abs_LFC']]) & min_lfc < avg_exps[['abs_LFC']], , drop = F])
    # report what we lose
    print(paste('LFC cutoff of', min_lfc, 'left', length(genes_lfc), 'of', length(rownames(seurat_object_cell_type))))
    # subset to those genes
    seurat_object_cell_type <- seurat_object_cell_type[genes_lfc, ]
    # convert to sca
    sca_ct <- seurat_object_to_singlecellassay(seurat_object_cell_type, assay=assay)

    # filter to the min.pct
    expressed_genes <- freq(sca_ct) > freq_expressed
    sca_ct <- sca_ct[expressed_genes, ]

    # do the analysis
    mast_result <- perform_MAST(sca_ct, contrast_var=contrast_var, fixed_effects=fixed_effects, random_effects=random_effects, nagq0 = nagq0)
    # add the genes present in the assay
    mast_result[['all_genes']] <- cell_type_genes
    # get a safe file name
    cell_type_safe <- gsub(' |/', '_', cell_type)
    cell_type_safe <- gsub('-', '_negative', cell_type_safe)
    cell_type_safe <- gsub('\\+', '_positive', cell_type_safe)
    cell_type_safe <- gsub('\\)', '', cell_type_safe)
    cell_type_safe <- gsub('\\(', '', cell_type_safe)
    # write a rds if requested
    if(!is.null(rds_loc)){
      # get the full path
      rds_full_path <- paste(rds_loc, contrast_var, '_', cell_type_safe, '.rds', sep = '')
      # write the result
      saveRDS(mast_result, rds_full_path)
    }
    # write a tsv if requested
    if(!is.null(tsv_loc)){
      # write a result per method
      for(method in names(mast_result)){
        # add the bonferoni correction
        mast_result[[method]][['dt']][['bonferroni']] <- mast_result[[method]][['dt']][['Pr(>Chisq)']] * length(cell_type_genes)
        # a p of larger than 1 is impossible
        mast_result[[method]][['dt']][!is.na(mast_result[[method]][['dt']][['bonferroni']]) &
                                        mast_result[[method]][['dt']][['bonferroni']] > 1, 'bonferroni'] <- 1
        # paste the full path
        tsv_full_path <- paste(tsv_loc, contrast_var, '_', cell_type_safe, '_', method, '.tsv', sep = '')
        # merge into one file containing the LFC and the P
        mast_result_wlfc <- merge(mast_result[[method]][['dt']], mast_result[[method]][['lfc']][, c('primerid', 'contrast', 'logFC', 'varLogFC')], by = c('primerid', 'contrast'))
        # write the result
        write.table(mast_result_wlfc, tsv_full_path, sep = '\t', row.names = F, col.names = T)
        # simplify if requested
        if(simplify_result){
          mast_result_celltype_simplified <- mast_output_dt_to_hurdle_summary(mast_result[[method]][['dt']])
          # merge with LFC
          mast_result_celltype_simplified <- merge(mast_result_celltype_simplified, mast_result[[method]][['lfc']][, c('primerid', 'contrast', 'logFC', 'varLogFC')], by = c('primerid', 'contrast'))
          tsv_simplified_full_path <- paste(tsv_loc, contrast_var, '_', cell_type_safe, '_', method, '_simplified.tsv', sep = '')
          # and save the result
          write.table(mast_result_celltype_simplified, tsv_simplified_full_path, sep = '\t', row.names = F, col.names = T)
        }
        # write the formula we used
        formula_used <- data.frame(capture.output(print(mast_result[[method]][['zlm']])))[2,1]
        # set the output loc
        formula_full_path <- paste(tsv_loc, contrast_var, '_', cell_type_safe, '_', method, '.formula', sep = '')
        # save the result
        write.table(print(formula_used), formula_full_path, row.names = F, col.names = F)
      }
    }
  }
  return(mast_result)
}

# perform regression for each cell type
glmer_per_cell_type <- function(seurat_object, cell_types=NULL, cell_type_column='cell_type', contrast_var='peak_ck_mb', fixed_effects=c('cngeneson','gender','age', 'timepoint.final'), random_effects=c('assignment.final','lane'), nagq0=F, assay='SCT', freq_expressed=0.1, tsv_loc=NULL, rds_loc=NULL, simplify_result=F, min_lfc=NULL){
  # setup result per cell_type
  result_per_cell_type <- list()
  # check which cell types to use
  cell_types_to_use <- unique(seurat_object@meta.data[[cell_type_column]])
  # use a smaller set if requested
  if(!is.null(cell_types)){
    cell_types_to_use <- intersect(cell_types_to_use, cell_types)
  }
  # remove NA ones
  cell_types_to_use <- cell_types_to_use[!is.na(cell_types_to_use)]
  # loop the cell types
  result_per_cell_type <- foreach(cell_type = cell_types_to_use) %dopar% {
    result_cell_type <- do_mast_celltype(seurat_object,
                                         cell_type, 
                                         cell_type_column = cell_type_column, 
                                         contrast_var = contrast_var, 
                                         fixed_effects = fixed_effects,
                                         random_effects = random_effects, 
                                         nagq0 = nagq0,
                                         assay = assay,
                                         freq_expressed = freq_expressed, 
                                         tsv_loc = tsv_loc, 
                                         rds_loc = rds_loc, 
                                         simplify_result = simplify_result, 
                                         min_lfc = min_lfc)
  }
  # set the names
  names(result_per_cell_type) <- cell_types_to_use
  return(result_per_cell_type)
}

# create a simplified version of the MAST output
mast_output_dt_to_hurdle_summary <- function(mast_dt){
  # grab just the hurdles
  mast_dt_hurdle <- mast_dt[mast_dt[['component']] == 'H', c('primerid', 'Pr(>Chisq)'), with = F]
  # grab the LFCs
  mast_dt_lfc <- mast_dt[mast_dt[['component']] == 'logFC', c('primerid', 'contrast', 'coef', 'ci.hi', 'ci.lo', 'z'), with = F]
  # merge together
  mast_dt_simplified <- merge.data.table(mast_dt_hurdle, mast_dt_lfc, by = 'primerid', all.x = T, all.y = T)
  return(mast_dt_simplified)
}


```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
####################
# Main Code        #
####################

# object locations
object_loc <- paste('/groups/umcg-franke-scrna/tmp01/releases/blokland-2020/v1/seurat/')
cardio_object_loc <- paste(object_loc, 'cardio.integrated.20210301.rds', sep = '')

# location of the clinical variables
clinvar_loc <- '/groups/umcg-franke-scrna/tmp01/releases/blokland-2020/v1/clinical_parameters/scRNAseq_clinical_data_20200910_final_notextfields.tsv'
inclusion_list_loc <- '/groups/umcg-franke-scrna/tmp01/releases/blokland-2020/v1/metadata/included_participants.txt'
# get the participants to include
inclusion_list <- read.table(inclusion_list_loc, header = F)$V1

# where to place the output
output_loc <- '/groups/umcg-franke-scrna/tmp01/releases/blokland-2020/v1/clinical_parameters/peakckmb_to_scrnaseq/cardio_integrated_20210301_all_timepoints/sct/'

```

```{r read_clinvars, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# read the clinvar table
clinvar <- read.table(clinvar_loc, header = T, sep = '\t', dec = ',', stringsAsFactors = F)
# calculate the peak ck_mb
clinvar <- add_peak_ck_mb(clinvar)
# add log of peak and baseline ckmb
clinvar$log_peak_ck_mb <- log(clinvar$peak_ck_mb)
clinvar$log_ck_mb <- log(clinvar$ck_mb)
# add the age
clinvar$age <- apply(clinvar, 1, function(x){
  age <- calc_age_in_days(as.character(x[['date_birth']]), split_character =  '/', year_col = 3, month_of_year_col = 2, day_of_month_col = 1)
  return(age)
})


```

```{r read_seurat, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# read the object
cardio.integrated <- readRDS(cardio_object_loc)

```

```{r subset_seurat, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# subset to the participants we are interested in (will also remove all HC)
cardio.integrated <- cardio.integrated[, cardio.integrated@meta.data[['assignment.final']] %in% inclusion_list]

```

```{r add_clinvars, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# add the relevant clinical variables
cardio.integrated <- add_clinical_variables(seurat_object = cardio.integrated, clinvar_object = clinvar, clinical_vars_to_add = c('peak_ck_mb', 'log_peak_ck_mb', 'age'), assignment_column_clinical_prepend = 'TEST_')
# make the age something of a more similiar scale to the expression
cardio.integrated@meta.data[['age']] <- cardio.integrated@meta.data[['age']] / 365.25
```

```{r add_clinvars, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

# do the analysis
result_per_celltype <- glmer_per_cell_type(seurat_object = cardio.integrated, cell_types = c('monocyte'), cell_type_column = 'cell_type_lowerres', contrast_var = 'peak_ck_mb', fixed_effects = c('cngeneson','gender','age', 'timepoint.final', 'chem'), random_effects = c('assignment.final','lane'), nagq0 = F, assay = 'SCT', freq_expressed = 0.1, tsv_loc = output_loc, rds_loc = output_loc, simplify_result = T, min_lfc = 0.1)

```
